src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -404,114 +383,93 @@ async def build_pex(
-            synthetic_lock = True
-            _fc = request.requirements.file_content
-            lock_path, lock_bytes = (_fc.path, _fc.content)
-            requirements_file_digest = await Get(Digest, CreateDigest([_fc]))
-
-        if _is_probably_pex_json_lockfile(lock_bytes):
-            should_resolve = False
-            header_delimiter = "//"
-            requirements_file_digest = await Get(
-                Digest,
-                CreateDigest(
-                    [FileContent(lock_path, _strip_comments_from_pex_json_lockfile(lock_bytes))]
-                ),
-            )
-            requirement_count = _pex_lockfile_requirement_count(lock_bytes)
-            argv.extend(["--lock", lock_path])
-            argv.extend(request.requirements.req_strings)
-        else:
-            header_delimiter = "#"
-            # Note: this is a very naive heuristic. It will overcount because entries often
-            # have >1 line due to `--hash`.
-            requirement_count = len(lock_bytes.decode().splitlines())
-            argv.extend(["--requirement", lock_path, "--no-transitive"])
-
-        if should_validate_metadata(request.requirements, python_setup):  # type: ignore[arg-type]
-            metadata = PythonLockfileMetadata.from_lockfile(
-                lock_bytes,
-                **(dict() if synthetic_lock else dict(lockfile_path=lock_path)),
-                resolve_name=resolve_name,
-                delimeter=header_delimiter,
-            )
-
+            argv.extend(["--requirement", lockfile.lockfile_path, "--no-transitive"])
+            is_network_resolve = True
+        if lockfile.metadata and request.requirements.complete_req_strings:
             validate_metadata(
-                metadata, request.interpreter_constraints, request.requirements, python_setup  # type: ignore[arg-type]
+                lockfile.metadata,
+                request.interpreter_constraints,
+                lockfile.original_lockfile,
+                request.requirements.complete_req_strings,
+                python_setup,
             )
-
     else:
-        assert isinstance(request.requirements, PexRequirements)
-        is_monolithic_resolve = request.requirements.is_all_constraints_resolve
-        requirement_count = len(request.requirements.req_strings)
-
-        if request.requirements.constraints_strings:
-            constraints_file = "__constraints.txt"
-            constaints_content = "\n".join(request.requirements.constraints_strings)
-            constraints_file_digest = await Get(
-                Digest,
-                CreateDigest([FileContent(constraints_file, constaints_content.encode())]),
-            )
-            argv.extend(["--constraints", constraints_file])
-
+        # TODO: This is not the best heuristic for available concurrency, since the
+        # requirements almost certainly have transitive deps which also need building, but it
+        # is better than using something hardcoded.
```

Comment:
This is out of place now

Reply:
Moving here was intentional, because the "entire lockfile" case now uses a lockfile-specific implementation, which is capable of being completely accurate. In particular: the transitivity bit doesn't apply in a lockfile, because the lockfiles are actually transitive, and have a fairly precise count of deps. The PEX lockfile parsing code has its own comment about accuracy.